src/pstl_offload.cpp

```diff
@@ -48,43 +48,111 @@ __get_original_msize()
     return __orig_msize;
 }
 
+struct __hash_aligned_ptr
+{
+    uintptr_t operator()(void *p) const
+    {
+        // we know that addresses are at least page-aligned, so drop 11 right bits that are zeros,
+        // and multiply to some prime number for bits hashing
+        // TODO: should well-known hashing function be used?
+        constexpr unsigned shift = 11;
+        return ((uintptr_t)p >> shift) * 1111111111111111111LLU;
+    }
+};
+
+struct __large_aligned_ptr_header
+{
+    sycl::device* _M_device;
+    std::size_t   _M_size;
+};
+
+std::mutex __large_aligned_ptrs_mtx;
+// map user pointer to header
+std::unordered_map<void*, __large_aligned_ptr_header, __hash_aligned_ptr> __large_aligned_ptrs;
```

Comment:
When I prototyped the same idea several months ago, I faced an issue (as far as I remember on LULESH) when `free` was called after destroying the pointers table and I have resolved it by introducing the special "is alive" flag. Did you face something similar?

Reply:
Good point. It's unsafe to touch an objects with non-trivial dtor unconditionally.